projectq/backends/_sim/_classical_simulator.py

```diff
@@ -0,0 +1,165 @@
+
+        Returns:
+            int: The mask.
+        """
+        t = 0
+        for q in qureg:
+            t |= 1 << self._bit_positions[q.id]
+        return t
+
+    def read_register(self, qureg):
+        """
+        Reads a group of bits as a little-endian integer.
+
+        Args:
+            qureg (projectq.types.Qureg):
+                The group of bits to read, in little-endian order.
+
+        Returns:
+            int: Little-endian register value.
+        """
+        t = 0
+        for i in range(len(qureg)):
+            t |= self.read_bit(qureg[i]) << i
+        return t
+
+    def write_register(self, qureg, value):
+        """
+        Sets a group of bits to store a little-endian integer value.
+
+        Args:
+            qureg (projectq.types.Qureg):
+                The bits to write, in little-endian order.
+            value (int): The integer value to store. Must fit in the register.
+        """
+        if value < 0 or value >= 1 << len(qureg):
+            raise ValueError("Value won't fit in register.")
+        for i in range(len(qureg)):
+            self.write_bit(qureg[i], (value >> i) & 1)
+
+    def is_available(self, cmd):
+        return (cmd.gate == Measure or
+                cmd.gate == Allocate or
+                cmd.gate == Deallocate or
+                isinstance(cmd.gate, BasicMathGate) or
+                isinstance(cmd.gate, FlushGate) or
+                isinstance(cmd.gate, XGate))
+
+    def receive(self, command_list):
+        for cmd in command_list:
+            self._handle(cmd)
+        if not self.is_last_engine:
+            self.send(command_list)
+
+    def _handle(self, cmd):
+        if cmd.gate == Measure or isinstance(cmd.gate, FlushGate):
+            return
+
+        if cmd.gate == Allocate:
+            new_id = cmd.qubits[0][0].id
+            self._bit_positions[new_id] = len(self._bit_positions)
+            return
+
+        if cmd.gate == Deallocate:
+            old_id = cmd.qubits[0][0].id
+            pos = self._bit_positions[old_id]
+            low = (1 << pos) - 1
+            self._state = (self._state & low) | ((self._state >> 1) & ~low)
+            self._bit_positions = {
+                k: b - (0 if b < pos else 1)
+                for k, b in self._bit_positions.items()
+            }
+            return
+
+        controls_mask = self._mask(cmd.control_qubits)
+        meets_controls = self._state & controls_mask == controls_mask
+
+        if isinstance(cmd.gate, XGate):
+            assert len(cmd.qubits) == 1 and len(cmd.qubits[0]) == 1
+            target = cmd.qubits[0][0]
+            if meets_controls:
+                self.write_bit(target, not self.read_bit(target))
+            return
+
+        if isinstance(cmd.gate, BasicMathGate):
+            if meets_controls:
+                ins = [self.read_register(reg) for reg in cmd.qubits]
+                outs = cmd.gate.get_math_function(None)(ins)
```

Comment:
get_math_function needs cmd.qubits (if users override its behavior such as in the AdvancedBasicMath gate you proposed).